### ISSUE TITLE
Proposal: Integrate trusted cloud-native registry Harbor with Dragonfly to provide a joint image management and distribution solution to support containerized environments

### ISSUE THREAD

**Issue by steven-zou** [alibaba/Dragonfly#108]
**Integrate trusted cloud-native registry Harbor with Dragonfly to provide a joint image management and distribution solution to support containerized environments.**

### Backgrounds:
**Harbor:**  Project Harbor is an open source trusted cloud-native registry project that stores, signs, and scans content. Harbor extends the open source Docker Distribution by adding the functionalities usually required by users such as security, identity, and management. Having a registry closer to the build and run environment can improve the image transfer efficiency. Harbor supports replication of images between registries and also offers advanced security features such as user management, access control, and activity auditing. For more details, please refer to [README](https://github.com/vmware/harbor/blob/master/README.md).

**Dragonfly:** Dragonfly is an intelligent P2P based file distribution system. It aims to resolve issues related to low-efficiency, low-success rate and a waste of network bandwidth in file transferring process. Especially in large-scale file distribution scenarios such as application distribution, cache distribution, log distribution, **image distribution**, etc. For more details, please refer to [README](https://github.com/alibaba/Dragonfly/blob/master/README.md)

### Motivations:
With the emergence and development of Kubernetes, it's becoming possible to run and operate large-scale containerized applications and services in enterprise environments. Meanwhile, there are still existing big challenges which cannot be ignored. How to securely and effectively manage the lots of container images produced in the enterprise organizations and distribute them to the large-scale runtimes with less time and efforts when starting applications or services on demand. To address the above challenge, we should build a joint solution from the open source trust cloud-native registry **Harbor** and the open source intelligent P2P based file distribution system **Dragonfly**. 

**These two open sourced projects have very obviously complementary advantages to each other and the joint solution will definitely expand the scenarios of image lifecycle management and improve the securities, reliabilities, and efficiencies.**

**Comment by lowzj** [comment]
Very looking forward to the integration of Dragonfly and Harbor. 

In order to complete the task of publishing images to the `SuperNode`, Dragonfly's internal workflow is:
1. gets the publishing image task from `Image Distribution Driver`
2. gets the URLs of all layers of the image from container image registry, `Harbor Registry`
3. sends tasks to all `SuperNode`s of Dragonfly to pre-download the image's layers, and periodically records the task status for query
4. then if anyone want to pull that image, Dragonfly will help him to download the image by layer via P2P network established by `SuperNode`

So there are some features need to be completed in Dragonfly:
* Authentication mechanism ensures that the caller is trusted
* API set provides the capabilities to publish images to the `SuperNode` of Dragonfly
* `DFDaemon`(the proxy deployed on every machines) supports pulling private container images

And something need to be confirmed:
* which API provided by harbor that Dragonfly can use it to get URLs of all layers of the image
 * I think the `Image Distribution Driver` is the connector between container image registry and file distribution system. IMOP, it's better to use pub-sub pattern and to be an independent component. In this way, we can easily reach a very loosely coupling.

**Comment by lowzj** [comment]
I drew a new architecture design graph to add some new components of dragonfly need to be implemented.
![image](https://user-images.githubusercontent.com/2419214/43308506-6426de16-91b4-11e8-8ad9-fc8eac3cfb87.png)
* **Dragonfly Controller** provides required API for harbor to publish images. It also need to implement authentication/rate limiting mechanism to ensure security.
* **Heater** includes `image heater` and `file heater` that providing the ability to warm-up container images and files. It analyzes the layers of images and downloads them from harbor registry to the SuperNode local disk.

**Comment by steven-zou** [comment]
I think the following diagram is also a good reference for image distribution:
![image distribution](https://github.com/alibaba/Dragonfly/blob/master/docs/images/dfget-combine-container.png)

**Comment by lowzj** [comment]
> **08.15** the specification of API between Dragonfly and Harbor 

document: https://github.com/alibaba/Dragonfly/blob/master/docs/en/preheat.md

**Comment by steven-zou** [comment]
@lowzj 

About [API](https://github.com/alibaba/Dragonfly/blob/master/docs/en/preheat.md#post-apipreheat) `/api/preheat`, could we make it compatible with registry API?  Harbor is a registry, that means a client can use the standard docker client or registry API to get the image content.

Seems the designed API needs a new Harbor API to do that. If I'm mistaken, please correct me.

**Comment by lowzj** [comment]
@steven-zou 

There is no need to create a new Harbor API. Firstly, Dragonfly assembles the minifest url according to the registry API spec and param `url` of `/api/preheat`, and fetches minifest to get all the urls of image layers. Then Dragonfly downloads all the layers from Harbor.

**Comment by steven-zou** [comment]
@lowzj 

So, I think the following example should be ok. Why is the header an array? Why not use a map?

```
{
  "type": "image",
  "url": "https://<harbor_hostname>/v2/library/redis/manifests/latest",
  "header": ["Authorization: Bearer <TOKEN>"]
}
```

**Comment by steven-zou** [comment]
@lowzj 

Ok, got it. But please be aware that the `<image_name>` in harbor has a prefix of a project name like `library/redis`. You need to take care of that.

**Comment by gitzl** [comment]
请问 Dragonfly集成Harbour 支持https 现在做了那些工作？如我们想修改源码支持https  需要做那些工作和注意的地方?  请解答下。非常感激

**Comment by gitzl** [comment]
Does Dragonfly integrate with Harbour support HTTPS and do the work now? If we want to modify the source code support HTTPS need to do those work and attention?  Very grateful

**Comment by perriea** [comment]
An update on this issue ?

**Comment by allencloud** [comment]
> An update on this issue ?

It works with Dragonfly since Dragonfly has already supported the preheat API.  @perriea 

While we have worked out a demo for integration of Harbor and Dragonfly. But I am not sure if the Harbor side has made a plan to release the work. @steven-zou

**Comment by datavisoryushuzhang** [comment]
Does the preheat API support private Harbor images? (which need docker login)

**Comment by allencloud** [comment]
> Does the preheat API support private Harbor images? (which need docker login)

Yes, it can. You can add the login credentials in the headers. @datavisoryushuzhang

**Comment by jessiezhou0424** [comment]
when will dragonfly+harbor demo release, can't wait to use it!
